cmd/object-handlers.go

```diff
@@ -302,46 +299,33 @@ func (api objectAPIHandlers) CopyObjectHandler(w http.ResponseWriter, r *http.Re
 		return
 	}
 
-	// Size of object.
-	size := objInfo.Size
-
-	pipeReader, pipeWriter := io.Pipe()
-	go func() {
-		startOffset := int64(0) // Read the whole file.
-		// Get the object.
-		gErr := objectAPI.GetObject(sourceBucket, sourceObject, startOffset, size, pipeWriter)
-		if gErr != nil {
-			errorIf(gErr, "Unable to read an object.")
-			pipeWriter.CloseWithError(gErr)
+	srcMetadata := objInfo.UserDefined
+	updateMetadata := isMetadataUpdate(srcMetadata, newMetadata)
+	if updateMetadata {
```

Comment:
if `updateMetadata` is true and dst != src, then lock is not held on dst during copyobject